apps/vmq_diversity/test/cache_test.lua

```diff
@@ -0,0 +1,121 @@
+mountpoint = ""
+client_id = "test-client"
+user = "test-user"
+acls = {{pattern = "a/+/c"}, 
+        {pattern = "c/#"}}
+
+-- same rules for publish and subscribe
+assert(auth_cache.insert(mountpoint, client_id, user, acls, acls) == true)
+
+assert(auth_cache.match_subscribe(mountpoint, client_id, "a/+/c", 1) == true)
+assert(auth_cache.match_subscribe(mountpoint, client_id, "a/b/c", 1) == true)
+assert(auth_cache.match_subscribe(mountpoint, client_id, "c/b/a", 1) == true)
+assert(auth_cache.match_subscribe(mountpoint, client_id, "a/b/d", 1) == false)
+assert(auth_cache.match_subscribe(mountpoint, client_id, "d/d/d", 1) == false)
+
+assert(auth_cache.match_publish(mountpoint, client_id, "a/b/c", 1, "hello", false) == true)
+assert(auth_cache.match_publish(mountpoint, client_id, "c/b/a", 1, "hello", false) == true)
+assert(auth_cache.match_publish(mountpoint, client_id, "a/b/d", 1, "hello", false) == false)
+assert(auth_cache.match_publish(mountpoint, client_id, "d/d/d", 1, "hello", false) == false)
+
+-- complex ACLs
+complex_client_id = "test-client-complex"
+complex_acls = {
+    {
+        pattern = "a/+/c",
+        max_qos = 1,
+        max_payload_size = 5,
+        allowed_retain = false
+    }
+}
+assert(auth_cache.insert(mountpoint, complex_client_id, user, complex_acls, complex_acls) == true)
+assert(auth_cache.match_subscribe(mountpoint, complex_client_id, "a/+/c", 2) == false) -- not allowed due to max_qos 
+assert(auth_cache.match_subscribe(mountpoint, complex_client_id, "a/+/c", 1) == true) -- allowed
+
+assert(auth_cache.match_publish(mountpoint, complex_client_id, "a/b/c", 2, "hello", false) == false) -- not allowed due to max_qos
+assert(auth_cache.match_publish(mountpoint, complex_client_id, "a/b/c", 1, "hello world", false) == false) -- not allowed due to max_payload_size
+assert(auth_cache.match_publish(mountpoint, complex_client_id, "a/b/c", 1, "hello", true) == false) -- not allowed due to not allowed_retain
+assert(auth_cache.match_publish(mountpoint, complex_client_id, "a/b/c", 1, "hello", false) == true) -- allowed
+
+
+-- complex ACLs including modifiers
+modifiers_client_id = "test-client-modifiers"
+pub_modifiers_acls = {
+    {
+        pattern = "a/+/c",
+        modifiers = {
+            topic = "hello/world",
+            payload = "hello world",
+            qos = 1,
+            retain = true,
+            mountpoint = "override-mountpoint2"
+        }
+    }
+}
+sub_modifiers_acls = {
+    {
+        pattern = "a/+/c",
+        modifiers = {{"hello/world", 2}}
+    }
```

Comment:
I tried to test this out by adding this to the mongodb:

```json
db.vmq_acl_auth.insert({
    mountpoint: '', 
    client_id: 'test-client', 
    username: 'test-user',
    passhash: '$2a$10$uUR/uiro5U6MuBn2hbeeDuVxp8v0TOM1r6Wr5Y13SKNRcZXYkIJp.',
    publish_acl: [
        {pattern: 'a/b/c'}, 
        {pattern: 'a/+/d'}
    ], 
    subscribe_acl: [{
        'pattern': 'a/+/c',
        'max_qos': 2,
        'modifiers': [
            {
                'topic': 'new/topic/1',
                'qos': 1
            },
            {
                'topic': 'new/topic/2',
                'qos': 1
            }
        ]
    }]
})
```

but then I get:

```
2017-02-27 15:44:48.192 [error] <0.378.0>@vmq_diversity_plugin:check_modifiers:497 can't rewrite topic in auth_on_subscribe due to wrong format [{qos,1},{topic,<<"new/topic/1">>}]
2017-02-27 15:44:48.193 [error] <0.378.0>@vmq_diversity_cache:validate_modifiers:308 can't validate modifiers subscribe for [{1,[{<<"qos">>,1.0},{<<"topic">>,<<"new/topic/1">>}]},{2,[{<<"qos">>,1.0},{<<"topic">>,<<"new/topic/2">>}]}] ACL
```

in the logs. So my suggestion above is clearly wrong.